src/tests/ftest/util/apricot/apricot/test.py

```diff
@@ -1242,8 +1251,45 @@ def remove_temp_test_dir(self):
             errors.append("Error removing temporary test files")
         return errors
 
+    def report_timeout(self):
+        """Dump ULTs stacks if this test case was timed out."""
+        if not self._timeout_reported:
+            if self.timeout is not None and self.time_elapsed > self.timeout:
+                # dump engines ULT stacks upon test timeout
+                if self.dumped_engines_stacks is False:
+                    self.dumped_engines_stacks = True
+                    self.log.info("Test status has timed-out, dumping ULT stacks")
+                    dump_engines_stacks(self.hostlist_servers)
+
+    def fail(self, message=None):
+        # dump engines ULT stacks upon test failure
+        if self.dumped_engines_stacks is False:
+            self.dumped_engines_stacks = True
+            self.log.info("Test has failed, dumping ULT stacks")
+            dump_engines_stacks(self.hostlist_servers)
+        super().fail(message)
+
+    def error(self, message=None):
+        # dump engines ULT stacks upon test error
+        if self.dumped_engines_stacks is False:
+            self.dumped_engines_stacks = True
+            self.log.info("Test has errored, dumping ULT stacks")
+            dump_engines_stacks(self.hostlist_servers)
+        super().error(message)
+
     def tearDown(self):
         """Tear down after each test case."""
+
+        # dump engines ULT stacks upon test failure
+        # XXX check of Avocado test status during teardown is presently useless
```

Comment:
(pylint-fixme) XXX check of Avocado test status during teardown is presently useless